reckon-core/src/main/groovy/org/ajoberstar/reckon/core/GitVcs.java

```diff
@@ -0,0 +1,239 @@
+    return Optional.of(getBaseVersion().getNormal());
+  }
+
+  @Override
+  public Optional<Version> getPreviousVersion() {
+    return Optional.of(getBaseVersion().getAny());
+  }
+
+  private BaseVersion getBaseVersion() {
+    try (RevWalk walk = new RevWalk(repo)) {
+      walk.setRetainBody(false);
+
+      ObjectId headObjectId = repo.getRefDatabase().getRef("HEAD").getObjectId();
+      RevCommit headCommit = walk.parseCommit(headObjectId);
+
+      Set<GitVersion> versions = new HashSet<>();
+
+      for (Ref ref : repo.getRefDatabase().getRefs(Constants.R_TAGS).values()) {
+        Ref tag = repo.peel(ref);
+        ObjectId objectId = tag.getPeeledObjectId();
+        RevCommit revCommit = walk.parseCommit(objectId);
+        Optional<GitVersion> maybeVersion =
+            tagParser.apply(tag).map(version -> new GitVersion(version, revCommit));
+
+        if (maybeVersion.isPresent()) {
+          GitVersion version = maybeVersion.get();
+          walk.reset();
+          walk.setRevFilter(RevFilter.MERGE_BASE);
+          walk.markStart(version.getRevCommit());
+          walk.markStart(headCommit);
+
+          RevCommit mergeBase = walk.next();
+
+          versions.add(version);
+          // if (!headCommit.equals(mergeBase)) {
+          //     Version tagged = version.getVersion();
+          //     Version normal = Version.forIntegers(tagged.getMajorVersion(), tagged.getMinorVersion(), tagged.getPatchVersion());
+          //     versions.add(new GitVersion(normal, mergeBase));
+          // }
+        } else {
+          logger.debug("Tag {} could not be parsed as a version.", tag.getName());
+        }
+      }
+
+      GitVersion any = findBase(walk, headCommit, versions.stream());
+      GitVersion normal =
+          findBase(
+              walk,
+              headCommit,
+              versions.stream().filter(ver -> ver.getVersion().getPreReleaseVersion().isEmpty()));
+      return new BaseVersion(
+          normal.getVersion(), any.getVersion(), headCommit.equals(any.getRevCommit()));
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  private GitVersion findBase(RevWalk walk, RevCommit head, Stream<GitVersion> versions) {
+    try {
+      walk.reset();
+      walk.setRevFilter(RevFilter.ALL);
+      walk.markStart(head);
+
+      Map<RevCommit, List<GitVersion>> versionsByRev =
+          versions.collect(Collectors.groupingBy(GitVersion::getRevCommit));
+
+      return StreamSupport.stream(walk.spliterator(), false)
+          .flatMap(
+              rev -> {
+                List<GitVersion> matches = versionsByRev.get(rev);
+                if (matches == null) {
+                  return Stream.empty();
+                } else {
+                  // Parents can't be "nearer". Exclude them to avoid extra walking.
+                  Arrays.stream(rev.getParents())
+                      .forEach(
+                          parent -> {
+                            try {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Extract this nested try block into a separate method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.ajoberstar.com/coding_rules#rule_key=squid%3AS1141)